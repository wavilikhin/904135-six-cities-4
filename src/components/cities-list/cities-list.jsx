import React, { memo } from 'react';
import PropTypes from 'prop-types';
import withActiveItem from '../../hocs/with-active-item/with-active-item.jsx';

let CitiesList = memo((props) => {
  const { currentCity, uniqueCities, onActiveItemChange } = props;

  return (
    <section className="locations container">
      <ul className="locations__list tabs__list">
        {uniqueCities.map((city, i) => {
          return (
            <li key={`${i}-${city}`} className="locations__item">
              <a
                onClick={(e) => {
                  e.preventDefault();
                  onActiveItemChange(city);
                }}
                className={`locations__item-link tabs__item ${
                  city === currentCity ? 'tabs__item--active' : ''
                }`}
                href="#"
              >
                <span>{city}</span>
              </a>
            </li>
          );
        })}
      </ul>
    </section>
  );
});

CitiesList.propTypes = {
  currentCity: PropTypes.string.isRequired,
  uniqueCities: PropTypes.arrayOf(PropTypes.string).isRequired,
  onActiveItemChange: PropTypes.func.isRequired,
};

export default withActiveItem(CitiesList, {
  stateUpdateRequired: true,
});

export { CitiesList };
