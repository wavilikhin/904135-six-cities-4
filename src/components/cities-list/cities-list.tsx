import * as React from 'react';

import { connect } from 'react-redux';
import { ActionCreator as StateActionCreator } from '../../reducer/state/state';
import { getUniqueCities } from '../../reducer/data/selectors';
import { getCity } from '../../reducer/state/selectors';
import { Dispatch } from 'redux';
import { AppStateType } from '../../reducer/reducer';
import { AppActionCreator } from '../../reducer/types';

type StateToPropsTypes = {
  currentCity: string;
  uniqueCities: string[];
};

type DispatchToPropsTypes = {
  handleCityChange: (city: string) => void;
};

type Props = StateToPropsTypes & DispatchToPropsTypes;

const CitiesList: React.FC<Props> = React.memo(
  ({ currentCity, uniqueCities, handleCityChange }) => {
    return (
      <section className="locations container">
        <ul className="locations__list tabs__list">
          {uniqueCities.map((city, i) => {
            return (
              <li key={`${i}-${city}`} className="locations__item">
                <a
                  onClick={(e) => {
                    e.preventDefault();
                    handleCityChange(city);
                  }}
                  className={`locations__item-link tabs__item ${
                    city === currentCity ? 'tabs__item--active' : ''
                  }`}
                  href="#"
                >
                  <span>{city}</span>
                </a>
              </li>
            );
          })}
        </ul>
      </section>
    );
  },
);

const mapStateToProps = (state: AppStateType) => ({
  currentCity: getCity(state),
  uniqueCities: getUniqueCities(state),
});

// FIXME: Dispatch type
const mapDispathcToProps = (dispatch: Dispatch<AppActionCreator>) => ({
  handleCityChange(city: string) {
    dispatch(StateActionCreator.changeCiy(city));
  },
});

export { CitiesList };

export default connect<
  StateToPropsTypes,
  DispatchToPropsTypes,
  {},
  AppStateType
>(
  mapStateToProps,
  mapDispathcToProps,
)(CitiesList);
