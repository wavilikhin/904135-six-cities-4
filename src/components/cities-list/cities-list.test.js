import React from 'react';
import renderer from 'react-test-renderer';
import { CitiesList } from './cities-list.jsx';

const uniqueCities = [
  'Amsterdam',
  'Dusseldorf',
  'Hamburg',
  'Paris',
  'Brusseles',
  'Cologne',
];
const city = 'Amsterdam';

it(`Should render cities list with 6 cities`, () => {
  const tree = renderer
    .create(
      <CitiesList
        uniqueCities={uniqueCities}
        currentCity={city}
        onActiveItemChange={() => {}}
      />,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
