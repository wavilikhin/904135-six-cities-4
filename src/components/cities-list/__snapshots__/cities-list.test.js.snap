// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitiesList component snapshot test Should render cities list with 6 cities 1`] = `
<section
  className="locations container"
>
  <ul
    className="locations__list tabs__list"
  >
    <li
      className="locations__item"
    >
      <a
        className="locations__item-link tabs__item tabs__item--active"
        href="#"
        onClick={[Function]}
      >
        <span>
          Amsterdam
        </span>
      </a>
    </li>
    <li
      className="locations__item"
    >
      <a
        className="locations__item-link tabs__item "
        href="#"
        onClick={[Function]}
      >
        <span>
          Dusseldorf
        </span>
      </a>
    </li>
    <li
      className="locations__item"
    >
      <a
        className="locations__item-link tabs__item "
        href="#"
        onClick={[Function]}
      >
        <span>
          Hamburg
        </span>
      </a>
    </li>
    <li
      className="locations__item"
    >
      <a
        className="locations__item-link tabs__item "
        href="#"
        onClick={[Function]}
      >
        <span>
          Paris
        </span>
      </a>
    </li>
    <li
      className="locations__item"
    >
      <a
        className="locations__item-link tabs__item "
        href="#"
        onClick={[Function]}
      >
        <span>
          Brusseles
        </span>
      </a>
    </li>
    <li
      className="locations__item"
    >
      <a
        className="locations__item-link tabs__item "
        href="#"
        onClick={[Function]}
      >
        <span>
          Cologne
        </span>
      </a>
    </li>
  </ul>
</section>
`;
